### ISSUE TITLE
[Question] how does webdataset internally deal with num_workers>1

### ISSUE THREAD

**Issue by maxwellzh** [webdataset/webdataset#160]
Hello, I'm new to webdataset and thanks for this great work.

The document of webdataset says that the `wds.WebDataset` is inherited from pytorch `IterableDataset` and I read the pytorch [doc](https://pytorch.org/docs/stable/data.html#torch.utils.data.IterableDataset), which reminds us to carefully deal with the dataset when `num_workers>1` with `worker_init_fn`, otherwise there would be duplicate data. But with `wds.WebDataset`, whatever the `num_workers` I set, the data is unique. Do I miss anything? Or does webdataset internally deal with that? If so, how does it get the worker info without `worker_init_fn`? Thanks.

Here is my code
```python3
with wds.TarWriter("dest.tar") as sink:
    for i in range(5):
        sink.write({
            '__key__': f"{i:03}",
            "input.npy": np.asarray(i, dtype=np.int32), 
            "output.npy": np.asarray(i+1, dtype=np.int32)
        })

dataset = wds.WebDataset('dest.tar').decode().to_tuple(
    "input.npy", "output.npy")

# it seems num_workers won't make the data duplicated
dataloader = DataLoader(dataset, batch_size=1, num_workers=2)
for in_seq, target in dataloader:
    print(in_seq)
```

**Comment by austinmw** [comment]
Hi, I also am curious how wds works when num_workers>1. I can create a new issue if this is too unrelated:

I noticed that with pip version, num_workers>1 still iterates through entire dataset, but with source install, the length of the iterator seems to be reduced proportionally to the increase in num_workers.

**Comment by tmbdev** [comment]
WebDataset is just a small convenience interface for DataPipeline. It only splits by worker, not by node. 

I recommend using the pipeline interface. 

By default, wds.shuffle will seed the rng by process id and time at the start of each epoch. If you use that before `split_by_worker` (as you did in your pipeline), each worker will end up with a random subset of shards, possibly overlapping.

If you use `split_by_worker` first, then deterministically, shards will get assigned to workers, and if you put wds.shuffle afterwards, they will get shuffled with an unpredictable seed on each epoch. That is what the WebDataset wrapper does by default.

You can use wds.detshuffle to get deterministic shuffling.

TODO: add testcases, documentation

**Comment by maxwellzh** [comment]
Thanks for the clarification!

**Comment by FFrse** [comment]
Hi, thanks for the answer, but why the offered example of pipeline usage in readme first use wds.shuffle and then wds.split_by_worker

`dataset = wds.DataPipeline(`
    `wds.SimpleShardList(url),`
    `# at this point we have an iterator over all the shards`
   `wds.shuffle(100),`
    `wds.split_by_worker,`
    `# at this point, we have an iterator over the shards assigned to each worker`
    `wds.tarfile_to_samples(),`
    `wds.shuffle(1000),`
    `wds.decode("torchrgb"),`
    `# at this point, we have an list of decompressed training samples from each shard in this worker in sequence`
    `get_patches,  # note that can put iterator->iterator functions into the pipeline directly`
    `wds.shuffle(10000),`
    `wds.to_tuple("big.jpg", "json"),`
    `wds.batched(16)`
`)`

`batch = next(iter(loader))`
`batch[0].shape, batch[1].shape`
